File: Cpp/46.cpp
// Given a collection of distinct integers, return all possible permutations.
//
// Example:
//
// Input: [1,2,3]
// Output:
// [
//   [1,2,3],
//   [1,3,2],
//   [2,1,3],
//   [2,3,1],
//   [3,1,2],
//   [3,2,1]
// ]

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        permute(res, nums, 0);
        return res;
    }

    void permute(vector<vector<int>>& res, vector<int>& nums, int pos) {
        if (pos >= nums.size()) {
            res.push_back(nums);
            return;
        }

        for (int i = pos; i < nums.size(); ++i) {
            swap(nums[i], nums[pos]);
            permute(res, nums, pos + 1);
            swap(nums[i], nums[pos]);
        }
    }
};

int main() {
    vector<int> test = {1, 2, 3, 4};
    Solution s;
    vector<vector<int>> res = s.permute(test);
    for (auto row : res) {
        cout << "[";
        for (auto elem : row) {
            cout << elem << ", ";
        }
        cout << "]" << endl;
    }

    return 0;
}

File: Python/46.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        self.dfs(nums, [], res)
        return res
    
    def dfs(self, nums, path, res):
        if not nums:
            res.append(path)
            return
        for i in range(len(nums)):
            self.dfs(nums[:i]+nums[i+1:], path+[nums[i]], res)




class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        self.permuteNums(nums, 0, res)
        return res

    def permuteNums(self, nums, pos, res):
        if pos >= len(nums):
            res.append(nums[:])
            return

        for i in range(pos, len(nums)):
            nums[i], nums[pos] = nums[pos], nums[i]
            self.permuteNums(nums, pos+1, res)
            nums[i], nums[pos] = nums[pos], nums[i]

class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        c = Counter(nums)
        vals = sorted(c.keys())
        cnts = [c[v] for v in vals]
        res = [0]*len(nums)
        ans = []
        def dfs(vals, cnts, res, pos):
            if pos >= len(nums):
                ans.append(res[:])
                return

            for i in range(len(vals)):
                if cnts[i] == 0:
                    continue
                n = cnts[i]
                for j in range(n):
                    cnts[i] -=1
                    res[pos] = vals[i]
                    dfs(vals, cnts, res, pos+1)
                    cnts[i] +=1

        dfs(vals, cnts, res, 0)
        return ans
